cpp/src/arrow/util/bitmap_ops.cc

```diff
@@ -237,71 +237,105 @@ bool OptionalBitmapEquals(const std::shared_ptr<Buffer>& left, int64_t left_offs
 
 namespace {
 
-template <template <typename> class BitOp>
-void AlignedBitmapOp(const uint8_t* left, int64_t left_offset, const uint8_t* right,
-                     int64_t right_offset, uint8_t* out, int64_t out_offset,
-                     int64_t length) {
+template <template <typename> class BitOp, bool ComputeNewValidityCount>
+int64_t AlignedBitmapOp(const uint8_t* left, int64_t left_offset, const uint8_t* right,
+                        int64_t right_offset, uint8_t* out, int64_t out_offset,
+                        int64_t length) {
   BitOp<uint8_t> op;
   DCHECK_EQ(left_offset % 8, right_offset % 8);
   DCHECK_EQ(left_offset % 8, out_offset % 8);
 
+  uint8_t* outFront = out;
   const int64_t nbytes = BitUtil::BytesForBits(length + left_offset % 8);
   left += left_offset / 8;
   right += right_offset / 8;
   out += out_offset / 8;
+  uint64_t outPopCount = 0;
   for (int64_t i = 0; i < nbytes; ++i) {
     out[i] = op(left[i], right[i]);
   }
+  if (ComputeNewValidityCount) {
+    outPopCount = CountSetBits(outFront, out_offset, length);
+  }
+  return outPopCount;
 }
 
-template <template <typename> class BitOp>
-void UnalignedBitmapOp(const uint8_t* left, int64_t left_offset, const uint8_t* right,
-                       int64_t right_offset, uint8_t* out, int64_t out_offset,
-                       int64_t length) {
+template <template <typename> class BitOp, bool ComputeNewValidityCount>
```

Comment:
From your test result, looks pop count calculation doesn't impact performance.
If that's the case, I think we should remove this new template argument and always calc and return pop count in this function. Caller can simply ignore the return value if it don't need it.

Reply:
No, there is an performance difference, quite good I would think - for the aligned code-path. Hence, I added the template bool expecting the compiler to optimize out the validity count calculation when it is requested.

Reply:
But your benchmark graph shows no difference from unaligned op, and this function is only for unaligned op, no?

Reply:
Yes that is true. Was hoping this might shave off any additional performance in rare cases at the cost of one additional template argument. I can remove it.